Compile error in void setup() no matching function for call to 'esphome::Automation<int, const char*, const char*>::Automation(esphome::logger::LoggerMessageTrigger*&)'

### The problem

My esphome was working, after upgrade to 2025.7, I now get (on one of 5 esphome devices)  this error
```
Archiving .pioenvs/heizung/lib64d/libWiFi.a
Compiling .pioenvs/heizung/libbc6/Update/Updater.cpp.o
src/main.cpp: In function 'void setup()':
src/main.cpp:256:97: error: no matching function for call to 'esphome::Automation<int, const char*, const char*>::Automation(esphome::logger::LoggerMessageTrigger*&)'
  256 |   automation_id = new Automation<int, const char *, const char *>(logger_loggermessagetrigger_id);
      |                                                                                                 ^
In file included from src/esphome/components/binary_sensor/filter.h:3,
                 from src/esphome/components/binary_sensor/binary_sensor.h:5,
                 from src/esphome/core/application.h:26,
                 from src/esphome/components/api/api_frame_helper.h:17,
                 from src/esphome/components/api/api_connection.h:5,
                 from src/esphome.h:3,
                 from src/main.cpp:3:
src/esphome/core/automation.h:286:12: note: candidate: 'esphome::Automation<Ts>::Automation(esphome::Trigger<Ts ...>*) [with Ts = {int, const char*, const char*}]'
  286 |   explicit Automation(Trigger<Ts...> *trigger) : trigger_(trigger) { this->trigger_->set_automation_parent(this); }
      |            ^~~~~~~~~~
src/esphome/core/automation.h:286:39: note:   no known conversion for argument 1 from 'esphome::logger::LoggerMessageTrigger*' to 'esphome::Trigger<int, const char*, const char*>*'
  286 |   explicit Automation(Trigger<Ts...> *trigger) : trigger_(trigger) { this->trigger_->set_automation_parent(this); }
      |                       ~~~~~~~~~~~~~~~~^~~~~~~
src/esphome/core/automation.h:284:32: note: candidate: 'constexpr esphome::Automation<int, const char*, const char*>::Automation(const esphome::Automation<int, const char*, const char*>&)'
  284 | template<typename... Ts> class Automation {
      |                                ^~~~~~~~~~
src/esphome/core/automation.h:284:32: note:   no known conversion for argument 1 from 'esphome::logger::LoggerMessageTrigger*' to 'const esphome::Automation<int, const char*, const char*>&'
src/esphome/core/automation.h:284:32: note: candidate: 'constexpr esphome::Automation<int, const char*, const char*>::Automation(esphome::Automation<int, const char*, const char*>&&)'
src/esphome/core/automation.h:284:32: note:   no known conversion for argument 1 from 'esphome::logger::LoggerMessageTrigger*' to 'esphome::Automation<int, const char*, const char*>&&'
Compiling .pioenvs/heizung/lib942/ESPAsyncWebServer/AsyncEventSource.cpp.o
Compiling .pioenvs/heizung/lib942/ESPAsyncWebServer/AsyncJson.cpp.o
Compiling .pioenvs/heizung/lib942/ESPAsyncWebServer/AsyncMessagePack.cpp.o
Compiling .pioenvs/heizung/lib942/ESPAsyncWebServer/AsyncWebHeader.cpp.o
Compiling .pioenvs/heizung/lib942/ESPAsyncWebServer/AsyncWebServerRequest.cpp.o
Archiving .pioenvs/heizung/libbc6/libUpdate.a
Compiling .pioenvs/heizung/lib942/ESPAsyncWebServer/AsyncWebSocket.cpp.o
*** [.pioenvs/heizung/src/main.cpp.o] Error 1
```



### Which version of ESPHome has the issue?

2025.7.0

### What type of installation are you using?

Home Assistant Add-on

### What platform are you using?

ESP32

### Component causing the issue

_No response_

### YAML Config

```yaml
#
#     ┌───────────┐
#     │  HEIZUNG  │                                              ┌─Y─ Ground
#     |           |                                              ├─2─ Nullleiter───────────────   to USB GND
#     │    L,N,PE ├3─────────────────────────────────────────────┴─1─ Phase 230V─────────────┬─   to USB 5V
#     │      Heat ├────────────────────────────────────────────────4─ relais channel 1 ──────┤    Wärmeforderung 
#     │       Umw ├────────────────────────────────────────────────3─ relais channel 3 ──────┤    Umwälzen
#     │         N ├────────────────────────────────────┬──┐                                  │
#     └───╥───╥───┘    ╔══════════════════════════════════┴────────5─ relais channel 4 ──────┤    Boiler Ventil 
#    Rückl║   ║Vorl    ║    ┌────────┐──dallas_temp    │              0x307673201864ff28     │    Boiler Temperatur
#         ║   ╠═══════VNT═══│ BOILER │════════════╗    │                                     │
#         ║   ║             └─╥────╥─┘            ║    │                                     │
#         ║   ║               ╚ZIRK╝              ║    │N                                    │
#         ║   ║                 ╚════════════════ ║ ═══┴───────────6─ relais channel 2 ──────┘    Zirkulieren   
#         ║   ║                                   ║
#         ║   ║dalles_temp                        ║                   0x14b84d201864ff28          Vorlauf Heizungen
#         ║   ╠════════Heizungen Eingangsseite══╗ ║
#         ║   ╚════════Heizungen Gartenseite══╗ ║ ║
#         ║                                   ║ ║ ║
#         ╠════════════dallas_temp════════════╝ ║ ║                   0xda424c201864ff28          Rücklauf Garten
#         ╠════════════dallas_temp══════════════╝ ║                   0x3e2849201864ff28          Rücklauf Eingang
#         ╚════════════dallas_temp════════════════╝                   0x2e224e201864ff28          Rücklauf Wasser
#
# Warmwasser: Wird zwischen wasser_min und wasser_max gehalten, ventilgesteuert. Der Heizungsvorlauf 
#   Kann also viel wärmer sein, wenn gewünscht. 
#
# Heizung: Im Sommermodus ist sie aus, außer für Wasser. Im Wintermodus (= not Sommermodus) wird geheizt
#
# Heizperiode: In der Heizperiode wird jeden Abend geprüft, ob HASS house_summermode bereitstellt 
#   Dieser sagt, welcher Modus eingestellt werden soll summermode = house_summermode. 
#   Gilt von 01.10T00:00 bis 31.03T23:59


substitutions:
  delta_under_min_alarm:     "1.0" # if the boiler is this amount below the boiler min temperature, then trigger the alarm, that the heater probably has a problem. 
  delta_send_vorrucklauf:    "1.0" #  0.15 voerher Änderungen im Vor und Rücklauf sind in der Regel stark und plötzlich. Ein angleich an die Raumtemperatur ist expontiell. 
  delta_send_boiler:         "0.04" # Natural cooling is ~0.25K/h = 0.0046 K/minute - over the course of 8 hours normally. 0.075 normally gives a value every 16 minutes. So 32 data points normally. But this is cutting off the heating peak a bit too much
#ECO  eco_mode_armed_threshold:  "35.0" # °C 
#ECO  eco_mode_armed_hysteresis: "60minutes" # time

esphome:
  name: heizung
  friendly_name: Heizung
  includes: 
    - heizung.h  

esp32:
  board: esp32dev
  framework:
    type: arduino

logger:
  level: DEBUG
  on_message:
    level: ERROR
    then:
      - text_sensor.template.publish:
         id: error_text
         state: !lambda |-
            return message;

api:
  encryption:
    key: "asddsdfsadfas="

ota:
  - platform: esphome
    password: "afsdasdf"

wifi:
  ssid: !secret wifi_ssid
  password: !secret wifi_password
  ap:
    ssid: "Heizung Fallback Hotspot"
    password: !secret wifi_captive_password
captive_portal:

web_server:
  port: 80
  version: 3



i2c:
  sda: GPIO21
  scl: GPIO22
  scan: true

external_components:
  - source: github://pr#4684
    components: seeedmultichannelrelay 
    refresh: 15s

seeedmultichannelrelay:
  id: SeedMultiChannelRelay_ID
  address: 0x11 

one_wire:
  - platform: gpio
    pin: 4

switch:
  - id: heizen #  actually htis is a "request heat" signal to the heater, this is connected to the port, where normally, a room thermostat is connected to
    name: heizen 
    platform: seeedmultichannelrelay
    inverted: true
    channel: 1
    seeedmultichannelrelay_id: SeedMultiChannelRelay_ID 
    restore_mode: RESTORE_DEFAULT_ON
    icon: "mdi:heating-coil"

    on_turn_on:
      - number.set:
          id: vorlauf_max
          value: !lambda return id(vorlauf_heizungen).state;
      - number.set:
          id: rucklauf_max
          value: !lambda |-
            return std::max({
              id(rucklauf_heizung_eingang).state, 
              id(rucklauf_heizung_garten).state, 
              id(rucklauf_wasser).state
            });

    #web_server_sorting_weight: 10
    
  - id: zirkulieren # warm water circulation pump
    name: zirkulieren
    platform: seeedmultichannelrelay
    channel: 2
    inverted: true
    seeedmultichannelrelay_id: SeedMultiChannelRelay_ID
    restore_mode: RESTORE_DEFAULT_OFF
    icon: "mdi:water-pump"
    #web_server_sorting_weight: 12

  - id: umwalzen # warm water circulation pump
    name: umwälzen
    platform: seeedmultichannelrelay
    channel: 3
    inverted: true
    seeedmultichannelrelay_id: SeedMultiChannelRelay_ID
    restore_mode: RESTORE_DEFAULT_OFF
    icon: "mdi:water-pump"
    #web_server_sorting_weight: 11

  - id: rucklauf_wasser_ventil_switch # warm water circulation pump
    platform: seeedmultichannelrelay
    name: Rücklaufventil Warmwasserkreis
    disabled_by_default: true
    channel: 4
    inverted: true
    seeedmultichannelrelay_id: SeedMultiChannelRelay_ID
    restore_mode: RESTORE_DEFAULT_OFF
    icon: "mdi:water-pump"
    #web_server_sorting_weight: 13

  - id: summermode # of on, then some rules ar active, such as: switch off the heater wenn water is warm enough and switch on when too cold. T
    name: "Sommermodus"
    platform: template
    optimistic: true
    restore_mode: RESTORE_DEFAULT_OFF
    icon: "mdi:weather-sunny"
    #web_server_sorting_weight: 14
    on_turn_on:
      - switch.turn_off: heizen
      - switch.turn_off: umwalzen

    on_turn_off:
      - switch.turn_on: heizen
      - switch.turn_on: umwalzen
    entity_category: config      

  - id: purge_zirk_done
    platform: template
    optimistic: true
    restore_mode: ALWAYS_ON


  - id: error # report an error, currently only one reason: water is below min_temparature - 1°. That should probably only happen, when the heating does not kick in at all. 
    name: "Störung an Heizung"
    platform: template
    optimistic: true
    restore_mode: RESTORE_DEFAULT_OFF
    #web_server_sorting_weight: -1
    icon: "mdi:alert"
    entity_category: diagnostic 

valve:
  - platform: template
    id: rucklauf_wasser_ventil
    lambda: |-
      if (id(rucklauf_wasser_ventil_switch).state) {
        return VALVE_OPEN;
      } else {
        return VALVE_CLOSED;
      }
    #web_server_sorting_weight: 13

    name: Rücklaufventil Warmwasserkreis
    open_action:
      - switch.turn_on: rucklauf_wasser_ventil_switch
    close_action:
      - switch.turn_off: rucklauf_wasser_ventil_switch

sensor:
  - name: "Rücklauf Heizung Eingangsseite"
    platform: dallas_temp
    address: 0x3e2849201864ff28
    update_interval: 15s
    id: rucklauf_heizung_eingang
    unit_of_measurement: "°C"
    icon: "mdi:thermometer-plus"
    device_class: "temperature"
    state_class: "measurement"
    accuracy_decimals: 1
    #web_server_sorting_weight: 50
    filters: 
      - sliding_window_moving_average:
          window_size: 3
          send_every: 2
      - delta: ${delta_send_vorrucklauf}
      - throttle: 1minutes
 
  - name: "Rücklauf Heizung Gartenseite"
    #web_server_sorting_weight: 51
    id: rucklauf_heizung_garten
    platform: dallas_temp
    address: 0xda424c201864ff28
    update_interval: 15s
    unit_of_measurement: "°C"
    icon: "mdi:thermometer-plus"
    device_class: "temperature"
    state_class: "measurement"
    accuracy_decimals: 1
    filters:
      - calibrate_linear:
          - 23.94 -> 23.50 # calibarion to vorlauf heizung sensor im Wasserbad
          - 45.00 -> 44.38 # calibarion to vorlauf heizung sensor im Wasserbad
      - sliding_window_moving_average:
          window_size: 3
          send_every: 2
      - delta: ${delta_send_vorrucklauf}
      - throttle: 1minutes

  - name: "Rücklauf Wasser"
    #web_server_sorting_weight: 52
    platform: dallas_temp
    id: rucklauf_wasser
    address: 0x2e224e201864ff28
    update_interval: 15s
    unit_of_measurement: "°C"
    icon: "mdi:thermometer-plus"
    device_class: "temperature"
    state_class: "measurement"
    accuracy_decimals: 1
    filters:
      - calibrate_linear:
          - 23.62 -> 23.50 # calibarion to vorlauf heizung sensor im Wasserbad
          - 44.38 -> 44.38 # calibarion to vorlauf heizung sensor im Wasserbad
      - sliding_window_moving_average:
          window_size: 3
          send_every: 2
      - delta: ${delta_send_vorrucklauf}
      - throttle: 1minutes

  - name: "Vorlauf Heizung"
    #web_server_sorting_weight: 53
    platform: dallas_temp
    id: vorlauf_heizungen
    address: 0x14b84d201864ff28
    update_interval: 15s
    unit_of_measurement: "°C"
    icon: "mdi:thermometer-minus"
    device_class: "temperature"
    state_class: "measurement"
    accuracy_decimals: 1
    filters:
      - calibrate_linear:
          - 23.32 -> 23.50 # calibarion to vorlauf heizung sensor im Wasserbad
          - 44.25 -> 44.38 # calibarion to vorlauf heizung sensor im Wasserbad
      - sliding_window_moving_average:
          window_size: 3
          send_every: 2
      - delta: ${delta_send_vorrucklauf}
      - throttle: 1minutes

  - name: "Warmwasser Boiler"
    #web_server_sorting_weight: 49
    platform: dallas_temp
    id: wasser_boiler
    address: 0x307673201864ff28
    update_interval: 15s
    unit_of_measurement: "°C"
    icon: "mdi:bathtub-outline"
    device_class: "temperature"
    state_class: "measurement"
    accuracy_decimals: 2
    filters:
      - calibrate_linear:
          - 0 -> 0
          - 100 -> 100
          #- 23.38 -> 23.5  # calibarion to vorlauf heizung sensor im Wasserbad
          #- 44.31 -> 44.38 # calibarion to vorlauf heizung sensor im Wasserbad
          #- 57.23 -> 60.00 # calibrated by measuring kitchen outflow!
      - exponential_moving_average:  
          send_every: 6
      - delta: ${delta_send_boiler}
      # Thotteling is obsolete: Send_every 6 and 10s update throttels already to one minute updates. 
 
    on_value:
      - if: # Water warm enough, stop heating
          condition: 
            lambda: 'return id(wasser_boiler).state >= id(wasser_ziel_max).state;'
          then:
            - if:
                condition:
                  switch.is_on: summermode
                then:             
                  - switch.turn_off: heizen
                  - delay: 5min # restwämrme ausnutzen. 
            - valve.close: rucklauf_wasser_ventil
            - switch.turn_off: umwalzen
      - if: # Water 75% warm, during heating phase, run zirkpump a bit to stir up water
          condition: 
            and:
              - switch.is_off: purge_zirk_done
              - lambda: 'return id(wasser_boiler).state >= id(wasser_ziel_min).state+(0.75*(id(wasser_ziel_max).state-id(wasser_ziel_min).state));'
          then:
            - switch.turn_on: purge_zirk_done
            - switch.turn_on: zirkulieren
            - delay: 60seconds
            - switch.turn_off: zirkulieren
            - delay: 60seconds
            - switch.turn_on: zirkulieren
            - delay: 60seconds
            - switch.turn_off: zirkulieren
            - delay: 60seconds
            - switch.turn_on: zirkulieren
            - delay: 60seconds
            - switch.turn_off: zirkulieren
            
      - if: # Water too cold, start heating
          condition: 
            lambda: 'return id(wasser_boiler).state <= id(wasser_ziel_min).state;'
          then:
            - switch.turn_on: heizen            
            - valve.open: rucklauf_wasser_ventil                  
            - switch.turn_off: purge_zirk_done            
            - switch.turn_on: umwalzen
      - if: # water so cold, something went wrong. 
          condition:
            and:  
              - switch.is_off: error
              - lambda: 'return id(wasser_boiler).state <= (id(wasser_ziel_min).state - ${delta_under_min_alarm} );'            
          then:
            - switch.turn_on: error
            - logger.log:
                level: ERROR
                format: "Heizung heizt nicht: Temperatur im Boiler bei %.1f. Minimum ist auf %.1f eingestellt."
                args: [ 'id(wasser_boiler).state', 'id(wasser_ziel_min).state' ]
      - if:
          condition:
            and:
              - switch.is_on: error                  
              - lambda: 'return (id(wasser_boiler).state > id(wasser_ziel_min).state);'
          then:
            - switch.turn_off: error
            - logger.log:
                level: INFO
                format: "Fehler zurückgesetzt: Temperatur im Boiler bei %.1f. Minimum ist auf %.1f eingestellt."
                args: [ 'id(wasser_boiler).state', 'id(wasser_ziel_min).state' ]     

  - platform: template
    #web_server_sorting_weight: 500
    id: rucklauf_wasser_pct
    name: Auslastung Wasser 
    icon: mdi:percent
    accuracy_decimals: 0
    unit_of_measurement: "%"
    state_class: "measurement"
    update_interval: never
    filters:
      - delta: 0.5

  - platform: template
    #web_server_sorting_weight: 500
    id: rucklauf_heizung_eingang_pct
    name: Auslastung Heizung Eingangsseite
    accuracy_decimals: 0
    icon: mdi:percent    
    unit_of_measurement: "%"
    state_class: "measurement"
    update_interval: never    
    filters:
      - delta: 0.5

  - platform: template
    #web_server_sorting_weight: 500
    id: rucklauf_heizung_garten_pct
    accuracy_decimals: 0    
    name: Auslastung Heizung Gartenseite
    icon: mdi:percent    
    unit_of_measurement: "%"
    state_class: "measurement"
    update_interval: never    
    filters:
      - delta: 0.5


  - platform: copy
    source_id: wasser_boiler
    id: "wasser_boiler_derivative"
    unit_of_measurement: "°C/h"
    icon: mdi:chart-line
    name: "Warmwasseränderung Boiler"
    #web_server_sorting_weight: 500
    filters:
      - lambda: |-
          static float last_value = 0;
          static float last_time = 0;
          float time   = (float) millis();
          if ((last_time == 0)|( time <= last_time )){
            last_value = x;
            last_time  = time;
            return {};
          }                    
          float change = ( ( x - last_value ) / ( time - last_time ) ) * 1000 * 60 * 60;
          last_value = x;
          last_time  = time;
          return change;

  - platform: copy
    source_id: wasser_boiler_derivative 
    id: wasser_boiler_derivative_ltn
    icon: mdi:chart-line
    #web_server_sorting_weight: 500
    name: Warmwasseränderung (Langzeit) Boiler
    accuracy_decimals: 3
    filters:
      - lambda: |-
          if(x>= 0.0){return {};};
          if(x<=-0.5){return {};};
          return x;
      - sliding_window_moving_average: 
          window_size: 10
          send_every: 3
          send_first_at: 3



number: 
  - name: Wasser Maximaltemp
    id: wasser_ziel_max
    platform: template
    optimistic: true
    restore_value: true    
    min_value: 50
    max_value: 70
    step: 1
    initial_value: 53
    #web_server_sorting_weight: 200
    on_value:
        then:
          - if:
              condition:
                lambda: 'return id(wasser_ziel_max).state <= id(wasser_ziel_min).state;'
              then:
                - delay: 100ms # seems to be needed for home assistant messaging, otherwise HASS does not show the change... ? 
                - number.increment: wasser_ziel_max


  - name: Wasser Minimaltemp
    id: wasser_ziel_min
    platform: template 
    optimistic: true
    restore_value: true
    min_value: 50
    max_value: 60
    step: 0.5    
    initial_value: 50
    #web_server_sorting_weight: 201
    on_value:
        then:
          - if:
              condition:
                lambda: 'return id(wasser_ziel_max).state <= id(wasser_ziel_min).state;'
              then:
                - delay: 100ms # seems to be needed for home assistant messaging, otherwise HASS does not show the change... ? 
                - number.decrement: wasser_ziel_min


  - name: Rücklauf Max
    id: rucklauf_max
    initial_value: 50
    restore_value: true
    platform: template 
    optimistic: true
    step: 0.5
    min_value: 0
    max_value: 100
  - name: Vorlauf Max
    id: vorlauf_max
    initial_value: 50
    restore_value: true
    platform: template 
    optimistic: true
    step: 0.5
    min_value: 0
    max_value: 100

  - platform: homeassistant
    id: roomtemperature
    entity_id: sensor.heizraum_temperature
    name: Raumtemperatur
    disabled_by_default: true

binary_sensor:

  - platform: homeassistant
    id: summermode_house
    entity_id: input_boolean.haus_in_summermode
    name: Sommermodus (Haus)
    disabled_by_default: true
    


text_sensor:
  - platform: template
    name: "Störung an Heizung (text)"
    id: error_text
    icon: "mdi:bug"
    #web_server_sorting_weight: -2
    entity_category: diagnostic 





time:
  - platform: sntp
    on_time: 
      - cron: "0 0 0 1 4 *" 
        then:
          - switch.turn_on: summermode
      - cron: "0 0 8,10,12,14,16,18,20,22 * 1-3,10-12 *"
        then:
          - lambda: |-
              id(summermode).publish_state(id(summermode_house).state);


      - cron: "0 /30 * * * *"
        then:
          if: 
            condition:
              - switch.is_off: error                  
            then: 
              - text_sensor.template.publish:
                  id: error_text
                  state: "N/A"
      # NEW
      - cron: "0 /8 * * * *"
        then:
          if: 
            condition:
                lambda: 'return (id(rucklauf_wasser_ventil).position == VALVE_OPEN);'
            then: 
              - switch.turn_on: zirkulieren
              - delay: 60seconds
              - switch.turn_off: zirkulieren

      - cron: "0 /1 * * * *"
        then: 
          - lambda: "heizung_auslastung();"
          - number.set:
              id: vorlauf_max
              value: !lambda |-
                return std::max({
                  id(vorlauf_max).state,
                  id(vorlauf_heizungen).state
                });
          - number.set:
              id: rucklauf_max
              value: !lambda |-
                return std::max({
                  id(rucklauf_max).state,
                  id(rucklauf_heizung_eingang).state, 
                  id(rucklauf_heizung_garten).state, 
                  id(rucklauf_wasser).state
                });
          - logger.log:
                      level: INFO
                      format: "Rücklauf Max: %.1f"
                      args: [ 'id(rucklauf_max).state' ]
```

### Anything in the logs that might be useful for us?

```txt

```

### Additional information

_No response_

I have been getting the same error on both 2025.7 **and 2025.7.1** with all my builds and configurations. 

```
src/main.cpp: In function 'void setup()':
src/main.cpp:222:99: error: no matching function for call to 'esphome::Automation<int, const char*, const char*>::Automation(esphome::logger::LoggerMessageTrigger*&)'
  222 |   automation_id_2 = new Automation<int, const char *, const char *>(logger_loggermessagetrigger_id);
      |                                                                                                   ^
In file included from src/esphome/components/binary_sensor/filter.h:3,
                 from src/esphome/components/binary_sensor/binary_sensor.h:5,
                 from src/esphome/core/application.h:26,
                 from src/esphome/components/api/api_frame_helper.h:17,
                 from src/esphome/components/api/api_connection.h:5,
                 from src/esphome.h:3,
                 from src/main.cpp:3:
src/esphome/core/automation.h:286:12: note: candidate: 'esphome::Automation<Ts>::Automation(esphome::Trigger<Ts ...>*) [with Ts = {int, const char*, const char*}]'
  286 |   explicit Automation(Trigger<Ts...> *trigger) : trigger_(trigger) { this->trigger_->set_automation_parent(this); }
      |            ^~~~~~~~~~
src/esphome/core/automation.h:286:39: note:   no known conversion for argument 1 from 'esphome::logger::LoggerMessageTrigger*' to 'esphome::Trigger<int, const char*, const char*>*'
  286 |   explicit Automation(Trigger<Ts...> *trigger) : trigger_(trigger) { this->trigger_->set_automation_parent(this); }
      |                       ~~~~~~~~~~~~~~~~^~~~~~~
src/esphome/core/automation.h:284:32: note: candidate: 'constexpr esphome::Automation<int, const char*, const char*>::Automation(const esphome::Automation<int, const char*, const char*>&)'
  284 | template<typename... Ts> class Automation {
      |                                ^~~~~~~~~~
src/esphome/core/automation.h:284:32: note:   no known conversion for argument 1 from 'esphome::logger::LoggerMessageTrigger*' to 'const esphome::Automation<int, const char*, const char*>&'
src/esphome/core/automation.h:284:32: note: candidate: 'constexpr esphome::Automation<int, const char*, const char*>::Automation(esphome::Automation<int, const char*, const char*>&&)'
src/esphome/core/automation.h:284:32: note:   no known conversion for argument 1 from 'esphome::logger::LoggerMessageTrigger*' to 'esphome::Automation<int, const char*, const char*>&&'
```

Going to have to comment out things one-by-one to figure out whats causing it.

--- a/esphome/components/logger/__init__.py
+++ b/esphome/components/logger/__init__.py
@@ -193,7 +193,7 @@ def validate_local_no_higher_than_global(value):
 Logger = logger_ns.class_("Logger", cg.Component)
 LoggerMessageTrigger = logger_ns.class_(
     "LoggerMessageTrigger",
-    automation.Trigger.template(cg.int_, cg.const_char_ptr, cg.const_char_ptr),
+    automation.Trigger.template(cg.uint8, cg.const_char_ptr, cg.const_char_ptr),
 )
 
 
@@ -390,7 +390,7 @@ async def to_code(config):
         await automation.build_automation(
             trigger,
             [
-                (cg.int_, "level"),
+                (cg.uint8, "level"),
                 (cg.const_char_ptr, "tag"),
                 (cg.const_char_ptr, "message"),
             ],

--- a/tests/components/logger/test-on_message.host.yaml
+++ b/tests/components/logger/test-on_message.host.yaml
@@ -1,0 +1,18 @@
+logger:
+  id: logger_id
+  level: DEBUG
+  on_message:
+    - level: DEBUG
+      then:
+        - lambda: |-
+            ESP_LOGD("test", "Got message level %d: %s - %s", level, tag, message);
+    - level: WARN
+      then:
+        - lambda: |-
+            ESP_LOGW("test", "Warning level %d from %s", level, tag);
+    - level: ERROR
+      then:
+        - lambda: |-
+            // Test that level is uint8_t by using it in calculations
+            uint8_t adjusted_level = level + 1;
+            ESP_LOGE("test", "Error with adjusted level %d", adjusted_level);

